src/functions/org/apache/jmeter/functions/TimeShift.java

```diff
@@ -0,0 +1,186 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.apache.jmeter.functions;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.time.Year;
+import java.time.ZoneId;
+import java.time.ZoneOffset;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeFormatterBuilder;
+import java.time.format.DateTimeParseException;
+import java.time.temporal.ChronoField;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.jmeter.engine.util.CompoundVariable;
+import org.apache.jmeter.samplers.SampleResult;
+import org.apache.jmeter.samplers.Sampler;
+import org.apache.jmeter.threads.JMeterVariables;
+import org.apache.jmeter.util.JMeterUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+
+/**
+ * timeShifting Function permit to shift a date
+ *
+ * Parameters: - format date @see
+ * https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html
+ * (optional - defaults to epoch time in millisecond) - date to shift formated
+ * as first param (optional - defaults now) - amount of (seconds, minutes,
+ * hours, days or months ) to add (optional - default nothing is add ) -
+ * variable name ( optional )
+ *
+ * Returns: a formatted date with the specified number of (seconds, minutes,
+ * hours, days or months ) added. - value is also saved in the variable for
+ * later re-use.
+ *
+ * @since 3.3
+ */
+public class TimeShift extends AbstractFunction {
+    private static final Logger log = LoggerFactory.getLogger(TimeShift.class);
+
+    private static final String KEY = "__timeShift"; // $NON-NLS-1$
+
+    private static final List<String> desc = Arrays.asList(JMeterUtils.getResString("time_format_shift"),
+            JMeterUtils.getResString("date_to_shift"), JMeterUtils.getResString("value_to_shift"),
+            JMeterUtils.getResString("function_name_paropt"));
+
+    // Ensure that these are set, even if no paramters are provided
+    private String format = ""; //$NON-NLS-1$
+    private String dateToShift = ""; //$NON-NLS-1$
+    private String amountToShift = ""; //$NON-NLS-1$
+    private String variableName = ""; //$NON-NLS-1$
+    private ZoneId systemDefaultZoneID = ZoneId.systemDefault();
+
+    /** Date time format cache handler **/
+    private Cache<String, DateTimeFormatter> dateTimeFormatterCache = null;
```

Comment:
`DateTimeFormatter` is thread-safe. Does is make sense to make cache `static final` ?